src/ring_theory/fractional_ideal.lean

```diff
@@ -0,0 +1,417 @@
+  rcases I with ⟨I, aI, haI, hI⟩,
+  rcases J with ⟨I, aJ, haJ, hJ⟩,
+  use aI * aJ,
+  use mul_ne_zero haI haJ,
+  intros b hb,
+  apply submodule.mul_induction_on hb,
+  { intros m hm n hn,
+    obtain ⟨n', hn', n'_eq⟩ := hJ n hn,
+    rw [mul_smul, ←algebra.mul_smul_comm, ←n'_eq, mul_comm],
+    apply hI,
+    exact submodule.smul _ _ hm },
+  { rw [smul_zero],
+    apply is_integer_coe },
+  { intros x y hx hy,
+    rw [smul_add],
+    apply is_integer_add _ hx hy },
+  { intros r x hx,
+    rw [←mul_smul, mul_comm, mul_smul],
+    apply is_integer_smul _ hx },
+end
+
+instance : has_mul (fractional_ideal R S) := ⟨λ I J, ⟨I.1 * J.1, fractional_mul I J⟩⟩
+
+@[simp]
+lemma val_mul (I J : fractional_ideal R S) : (I * J).1 = I.1 * J.1 := rfl
+
+lemma mul_left_mono (I : fractional_ideal R S) : monotone ((*) I) :=
+λ J J' h, mul_le.mpr (λ x hx y hy, mul_mem_mul hx (h hy))
+
+lemma mul_right_mono (I : fractional_ideal R S) : monotone (λ J, J * I) :=
+λ J J' h, mul_le.mpr (λ x hx y hy, mul_mem_mul (h hx) hy)
+
+instance add_comm_monoid : add_comm_monoid (fractional_ideal R S) :=
+{ add_assoc := λ I J K, lattice.sup_assoc,
+  add_comm := λ I J, lattice.sup_comm,
+  add_zero := λ I, lattice.sup_bot_eq,
+  zero_add := λ I, lattice.bot_sup_eq,
+  ..fractional_ideal.has_zero,
+  ..fractional_ideal.has_add }
+
+instance comm_monoid : comm_monoid (fractional_ideal R S) :=
+{ mul_assoc := λ I J K, ext (submodule.mul_assoc _ _ _),
+  mul_comm := λ I J, ext (mul_comm _ _),
+  mul_one := λ I, begin
+    ext,
+    split; intro h,
+    { apply mul_le.mpr _ h,
+      rintros x hx y ⟨y', y'_mem_R, y'_eq_y⟩,
+      erw [←y'_eq_y, mul_comm, coe_mul_eq_smul],
+      exact submodule.smul _ _ hx },
+    { have : x * 1 ∈ (I * 1) := mul_mem_mul h (coe_mem_one _),
+      simpa }
+  end,
+  one_mul := λ I, begin
+    ext,
+    split; intro h,
+    { apply mul_le.mpr _ h,
+      rintros x ⟨x', x'_mem_R, x'_eq_x⟩ y hy,
+      erw [←x'_eq_x, coe_mul_eq_smul],
+      exact submodule.smul _ _ hy },
+    { have : 1 * x ∈ (1 * I) := mul_mem_mul (coe_mem_one _) h,
+      simpa }
+  end,
+  ..fractional_ideal.has_mul,
+  ..fractional_ideal.has_one }
+
+instance comm_semiring : comm_semiring (fractional_ideal R S) :=
+{ mul_zero := λ I, eq_zero_iff.mpr (λ x hx, submodule.mul_induction_on hx
+    (λ x hx y hy, by simp [mem_zero_iff.mp hy])
+    rfl
+    (λ x y hx hy, by simp [hx, hy])
+    (λ r x hx, by simp [hx])),
+  zero_mul := λ I, eq_zero_iff.mpr (λ x hx, submodule.mul_induction_on hx
+    (λ x hx y hy, by simp [mem_zero_iff.mp hx])
+    rfl
+    (λ x y hx hy, by simp [hx, hy])
+    (λ r x hx, by simp [hx])),
+  left_distrib := λ I J K, ext (mul_add _ _ _),
+  right_distrib := λ I J K, ext (add_mul _ _ _),
+  ..fractional_ideal.add_comm_monoid,
+  ..fractional_ideal.comm_monoid }
+end semiring
+
+section quotient
+
+/-! ### `quotient` section
+
+This section defines the ideal quotient of fractional ideals.
+
+In this section we need that each non-zero `y : R` has an inverse in
```

Comment:
Up till now, have you used that `R` is an integral domain? Or would `comm_ring` suffice? If so, I think it would make sense to only start assuming `R` is an integral domain from now on.

Reply:
The reason I chose `integral_domain` over `comm_ring` throughout the file is to ensure the definition of `I + J` and `I * J` works: if the denominators for `I` and `J` are `aI ≠ 0` and `aJ ≠ 0` respectively, then `I + J` and `I * J` have denominator `aI * aJ ≠ 0`. (And even Bourbaki only considers fractional ideals as submodules of the fraction field of an integral domain.)

I also considered taking `R` to be a `comm_ring` and have `a ∈ S` (where `S` is the submonoid at which we localize). That didn't work out when I started, but now after I've changed around some things it might work. Does that seem like the appropriate generalization?

Reply:
My only reason for weakening the assumption would be to make the life of the tc inference system easier. I guess in all applications `R` will be an integral domain. Giving the user an `a ∈ S` proof obligation doesn't sound ideal. So I would prefer to avoid that.

Reply:
OK, then I will keep it an integral domain and leave the generalization to `comm_ring` to someone else if they actually need it.